### DISCUSSION TITLE
Parsing a list with no brackets

### DISCUSSION THREAD

**Discussion by dalps** [c-cube/ocaml-containers#461] [Q&A] [answered]
I'm attempting this year's advent of code with OCaml and Containers. The first challenge requires parsing a list of pairs from a text file formatted like this:
```
3   4
4   3
2   5
1   3
3   9
3   3
```
While I could get the bracketless pair parser working, the list parser won't cope with the lack of brackets.
Here's my broken attempt - it just returns the empty list: 
```ocaml
let parse_int_pair_list =
  CCParse.U.(
    list ~start:"" ~stop:"" ~sep:""
      (pair ~start:"" ~stop:"" ~sep:"" int int))
```
So, is it possible to parse bracketless lists with CCParse?

**Comment by dalps** [comment]
Oh wait, I see why it returns the empty list immediately with those arguments... any hope of getting it to work without resorting to brackets?

**Comment by c-cube** [comment, marked as answer]
I have this alternative if you like:

```ocaml
# module P = CCParse
# let s = "1 2\n3 4\n5   6\n7  8";;
val s : string = "1 2\n3 4\n5   6\n7  8"
# P.parse_string P.(each_line (both (U.int <* many space) U.int)) s;;
- : ((int * int) list, string) result = Ok [(1, 2); (3, 4); (5, 6); (7, 8)]
```

If there are empty lines it's a bit more annoying (`each_line` is somewhat rigid) but you can get options and filter later with `CCList.keep_some`:

```ocaml
# let s = "1 2\n3 4\n5   6\n7  8\n";;
val s : string = "1 2\n3 4\n5   6\n7  8\n"
# P.parse_string P.(each_line (try_opt @@ both (U.int <* many space) U.int)) s;;
- : ((int * int) option list, string) result =
Ok [Some (1, 2); Some (3, 4); Some (5, 6); Some (7, 8); None]
```

**Reply by dalps** [reply]
Thank you!! Now I like the CCParse API even more.

The first alternative is exactly what I had in mind. I tried to see if it worked on my file - pretty sure there are no empty lines here:
```
3   4
4   3
2   5
1   3
3   9
3   3
```
Trying to parse it gives:
```ocaml
P.parse_file P.(each_line (both (U.int <* many space) U.int)) "input.txt";;
- : ((int * int) list, string) result =
Error "at line 6, char 1: expected non-empty sequence of chars for integer"
```
Have you tested it on a file? The second alternative works fine, so I could always use that.

**Reply by dalps** [reply]
Nevermind, I got it to work. Sorry for the noise.
```ocaml
# P.parse_file P.(many (both U.int U.int)) "input.txt";;
- : ((int * int) list, string) result =
Ok [(3, 4); (4, 3); (2, 5); (1, 3); (3, 9); (3, 3)]
```
